### ISSUE TITLE
StreamField Block groups sometimes render bizarrely 

### ISSUE THREAD

**Issue by coredumperror** [wagtail/wagtail#3704]
I have a Page model called FlexPage, which is defined in a site-specific app. It uses a custom set of StreamField Blocks, in addition to a set of blocks defined in the "core" app which are shared with other apps.

The core app has this code block:
```
base_flexpage_widgets = [
    ('grid_layout', ColumnStreamBlock(group="Layout")),
    ('full_html', RawHTMLBlock(label='Full HTML', group="Text")),
    ('rich_text', RichTextBlock(label='Rich Text', group="Text")),
    ('image_panel', ImagePanelBlock(label='Image Panel', group="Images")),
    ('image_link', ImageLinkBlock(label='Image Link', group="Images")),
    ('captioned_image', CaptionedImageBlock(label='Captioned Image', group="Images")),
    ('news_listing', NewsListingBlock(label='News Listing', group="Listing Blocks")),
    ('spacer', SpacerBlock(label='Spacer', group="Layout")),
]
```

And the site-specific app uses this definition for its body field:
```
    body = StreamField([
        ('two_column_layout', TwoColumnBlock(group="Layout")),
        ('three_column_layout', ThreeColumnBlock(group="Layout")),
        ('four_column_layout', FourColumnBlock(group="Layout")),
        ] + base_flexpage_widgets +  # Placed here to preserve ordering in admin dashboard
        [('event_listing', EventBlock(label='Event Listing', group="Listing Blocks")),
        (
            'campus_announcements_listing',
            CampusAnnouncementsBlock(label='Campus Announcements Listing', group="Listing Blocks")
        ),
        ('section_title', SectionTitleBlock(label='Section Title', group="Text")),
        ('menu_listing', MenuListingBlock(label='Menu Listing', group="Listing Blocks")),
        ('image_slider', ImageCarouselBlock(label='Hero Image Slider', group="Images"))
    ])
```

I would expect there to be four Groups in the SteamField widget display: "Images", "Layout", "Listing Blocks", and "Text". 

Instead, I get this:
![screenshot](https://user-images.githubusercontent.com/413571/28095605-26a421ee-6658-11e7-933d-3683473ad04d.png)

Oddly enough, this only seems to affect the StreamField widget list when it's shown _after_ an existing Block in the form. The list groups itself correctly with an empty block list, or when you press the "+" button that's above the first block. But the listing is busted if you press the + button that's below any existing block.

**Comment by gasman** [comment]
Fixed in #4303

**Comment by coredumperror** [comment]
Fantastic! Thanks!
